Add huggingface Galaxy data table schema reference (#3923)
galaxyproject/galaxy-hub: content/news/2026/2026-04-13-huggingface-data-table/index.md

---
title: "The huggingface Galaxy data table: schema and ontology"
date: "2026-04-13"
tease: "Galaxy is standardizing a shared data table for pre-downloaded HuggingFace models. This post is the reference for the 7-column schema, controlled vocabularies, and conventions."
tags: [tools, admin, ai]
subsites: [all]
autotoc: true
contributions:
  authorship:
    - arash77
---

A growing number of Galaxy tools need to present pre-downloaded
HuggingFace-compatible models to users at job submission time.
[Flux](https://github.com/bgruening/galaxytools/tree/master/tools/flux)
already uses this pattern, and future HuggingFace-backed tools can adopt the
same convention.  To avoid schema drift and duplicated administrator effort,
the Galaxy community is standardizing a [**shared Galaxy data table**](/admin/tools/data-tables/) named
`huggingface`.  This post is the **reference** for that schema.

**Key points:**

- The table has **7 columns**: `value`, `name`, `pipeline_tag`, `domain`, `free_tag`, `version`, `path`

> - It is recommended to **filter by `free_tag` + `version`** so different tools' rows never interfere

Reviewer: Since its a free tag, its hard to filter for a tool, because tool developers in general don't know what models are deployed. Would it not be better to recommend to tool authors to filter for "pipeline_tag" and/or "domain", and only use the free_tag as fallback when you need to narrow down your selection even more?

- New tools choose a **unique lowercase `free_tag`** as their namespace in the shared file

## The data table

The table is declared with this block in `tool_data_table_conf.xml`:

```xml
<!-- huggingface models -->
<table name="huggingface" comment_char="#" allow_duplicate_entries="False">
    <columns>value, name, pipeline_tag, domain, free_tag, version, path</columns>
    <file path="/opt/galaxy/tool-data/huggingface.loc" />
</table>
```

The sample file shipped with each tool (`tool-data/huggingface.loc.sample`)
uses the same 7-column layout.

## Column reference

| # | Column | Purpose |
|---|--------|---------|
| 0 | `value` | Unique row ID across the **whole table** |
| 1 | `name` | Human-readable label shown in the Galaxy select widget |
| 2 | `pipeline_tag` | Model role — see controlled vocabulary below |
| 3 | `domain` | Coarse data domain — see controlled vocabulary below |

> | 4 | `free_tag` | Per-tool-family tag; used as the primary XML filter |

Reviewer: is that really the primary filter? Or just now with the first tools?

Author: That was based on the Flux example which happens to filter by `free_tag`, but it's not the right general recommendation.

Author: Then we have to change the FLUX tool to match to the docs.

> | 5 | `version` | Tool version the row belongs to |

Reviewer: Is that really the tool version, or rather a model version, and a tool in certain versions can declare which model version that can consume.

Author: Correct, it should be the model version.

> | 6 | `path` | Path to the model data, a directory or a specific file |

Reviewer: ```suggestion
| 6 | `path` | Path to the model data, a directory or a specific file, depending on the model structure |
```

### `value` (column 0)

Must be **globally unique** across every row in `huggingface.loc`,
regardless of which tool added it.  Recommended pattern:

```

> <free_tag>_<version>_<slug>

Reviewer: Does huggingface provide a unique ID which we can use?

Author: Yes, we could use that.

```

Examples: `flux_1_dev`, `embedding_1.0_bge-small-en`,

> `mytool_2.1_base-model`.

Reviewer: a model should be consumable by multple tools. I would not put the toolname or anything here.

### `pipeline_tag` (column 2)

Use the **official HuggingFace pipeline tag** whenever one exists — browse
the full list at https://huggingface.co/models?pipeline_tag=…

Recommended values for this table include:

| Value | When to use |
|-------|-------------|
| `text-to-image` | Image generation models (Flux) |
| `automatic-speech-recognition` | ASR / transcription models |
| `feature-extraction` | Sentence / document embedding models |
| `tabular-classification` | Tabular ML classifiers |
| `tabular-regression` | Tabular ML regressors |
| `text-generation` | Causal / instruction-tuned LLMs |

Do **not** invent synonyms for existing HF tags (e.g. write
`automatic-speech-recognition`, not `asr`).

### `domain` (column 3)

A broad category for the type of data the model works with. For example:

`image` · `text` · `audio` · `tabular` · `sequence` · `video` · `multimodal`

### `free_tag` (column 4)

> A short, tool-family-specific identifier used as the primary filter in the

Reviewer: I would declare this as a fallback, if you need to narrow your model-space even more.

tool XML.  Currently registered public values:

| Value | Tool |
|-------|------|
| `flux` | Flux image-generation tool |

When adding a new HF-backed tool, choose a unique lowercase `free_tag` and
add it to this table when the tool is ready to be documented publicly.

### `version` (column 5)

> A version string used to filter rows in the tool XML.  Rows are only added

Reviewer: this is the model version I assume?

---

> title: "The huggingface Galaxy data table: schema and ontology"

Reviewer 2: ```suggestion
title: "The Hugging Face Galaxy data table: schema and ontology"
```

> title: "The Hugging Face Galaxy data table: schema and ontology"

Reviewer 3: Doesn't that belong in the Galaxy docs ?

Author: This post can serve as the announcement and link to the docs maybe latter?

Reviewer 3: that seems fine but i don't think we should add docs in the hub, we did that in the past and it just caused information to be scattered even more, with docs on the hub basically never getting updated

Author: Should I add it to https://github.com/galaxyproject/galaxy/blob/dev/doc/source/admin/data_tables.md?

Reviewer 3: That's fine, yes. You could also list the existing tables and document them in a followup.

date: "2026-04-13"

> tease: "Galaxy is standardizing a shared data table for pre-downloaded HuggingFace models. This post is the reference for the 7-column schema, controlled vocabularies, and conventions."

Reviewer 2: ```suggestion
tease: "Galaxy is standardizing a shared data table for pre-downloaded Hugging Face models. This post is the reference for the 7-column schema, controlled vocabularies, and conventions."
```

## The data table

The table is declared with this block in `tool_data_table_conf.xml`:

```xml
<!-- huggingface models -->
<table name="huggingface" comment_char="#" allow_duplicate_entries="False">
    <columns>value, name, pipeline_tag, domain, free_tag, version, path</columns>
    <file path="/opt/galaxy/tool-data/huggingface.loc" />
</table>
```

The sample file shipped with each tool (`tool-data/huggingface.loc.sample`)
uses the same 7-column layout.

## Column reference

| # | Column | Purpose |
|---|--------|---------|
| 0 | `value` | Unique row ID across the **whole table** |
| 1 | `name` | Human-readable label shown in the Galaxy select widget |
| 2 | `pipeline_tag` | Model role — see controlled vocabulary below |
| 3 | `domain` | Coarse data domain — see controlled vocabulary below |
| 4 | `free_tag` | Optional narrowing tag; fallback filter when `pipeline_tag`/`domain` alone are not specific enough |
| 5 | `version` | Model version |
| 6 | `path` | Path to the model data, a directory or a specific file, depending on the model structure |

### `value` (column 0)

Must be **globally unique** across every row in `huggingface.loc`,

> regardless of which tool added it.  HuggingFace already assigns unique model

Reviewer 2: ```suggestion
regardless of which tool added it.  Hugging Face already assigns unique model
```

IDs (in the form `<owner>/<model-name>`) — use those directly as the value
wherever possible, since they are stable and unambiguous.  If the same
underlying model is registered at more than one version, append the version:

```
<hf-model-id>
<hf-model-id>_<version>
```

Examples: `black-forest-labs/FLUX.1-dev`,
`sentence-transformers/all-MiniLM-L6-v2`,
`openai/whisper-large-v3_3.0`.

### `pipeline_tag` (column 2)

> Use the **official HuggingFace pipeline tag** whenever one exists — browse

Reviewer 2: ```suggestion
Use the **official Hugging Face pipeline tag** whenever one exists — browse
```

| `text-to-image` | Image generation models (Flux) |
| `automatic-speech-recognition` | ASR / transcription models |
| `feature-extraction` | Sentence / document embedding models |
| `tabular-classification` | Tabular ML classifiers |
| `tabular-regression` | Tabular ML regressors |
| `text-generation` | Causal / instruction-tuned LLMs |

Do **not** invent synonyms for existing HF tags (e.g. write
`automatic-speech-recognition`, not `asr`).

### `domain` (column 3)

A broad category for the type of data the model works with. For example:

`image` · `text` · `audio` · `tabular` · `sequence` · `video` · `multimodal`

### `free_tag` (column 4)

An optional short identifier used as a **fallback narrowing filter** in the
tool XML, for cases where `pipeline_tag` and `domain` alone are not specific
enough.  Because a model can be consumed by multiple tools, the `free_tag`
should not encode a specific tool name.  Currently registered public values:

| Value | Scope |
|-------|-------|
| `flux` | Flux image-generation models |

Only add a `free_tag` when you genuinely need to narrow the selection beyond
what `pipeline_tag`/`domain` provide.  Choose a short, lowercase, descriptive

> identifier and register it here when the tool is ready to be documented

Reviewer: here? Should the user update this blog post?

Reviewer 2: I think `here` should point to the `huggingface.loc.sample` file.

Author: > here? Should the user update this blog post?

Well ideally that would have been great to have a single source for them.

> huggingface.loc.sample

Since each tool has its own `.loc.sample`, there's no single data file to point to.

Author: Now we can have the galaxy docs as the source of truth.

publicly.

### `version` (column 5)

The **model version** string.  A tool declares in its XML which model version(s)
it can consume, allowing multiple versions of the same model to coexist in the
table.  Where possible, rows are only added, never removed or edited; when a new model version

> is deployed it gets a new row while old rows stay in place.

Reviewer 2: ```suggestion
is deployed it gets a new row while existing rows stay in place.
```

### `path` (column 6)

> The path to the model data on the server. It can be a directory (when the

Reviewer 2: ```suggestion
The path to the model data either on a local machine during tool development or on the Galaxy production server (maintained by admins). It can be a directory (when the
```

`.ckpt` checkpoint when the tool only needs that one file).

## XML filter convention

It is recommended that tool XML filters **primarily by `pipeline_tag` (column 2)
and/or `domain` (column 3)**, so only relevant model types are presented to the
user.  Add a `free_tag` or `version` filter only when you need to narrow the
selection further:

```xml
<param name="model" type="select" label="Model">
    <options from_data_table="huggingface">
        <filter type="static_value" column="2" value="<pipeline_tag>"/>
        <filter type="static_value" column="3" value="<domain>"/>
        <!-- optional: narrow further by version or free_tag -->
        <!-- <filter type="static_value" column="5" value="<version>"/> -->
        <!-- <filter type="static_value" column="4" value="<free_tag>"/> -->
    </options>
</param>
```

Example from the Flux tool (filters by `free_tag` to restrict to
Flux-specific model variants):

```xml
<options from_data_table="huggingface">
    <filter type="static_value" column="4" value="flux"/>
    <filter type="static_value" column="5" value="1"/>
</options>
```

## Complete example rows

> Each row in the `.loc` file is **TAB-separated** (7 fields). The comments show

Reviewer 2: ```suggestion
Each row in the `.loc` file is **TAB-separated** (7 columns). The comments show
```

tags: [tools, admin, ai]
subsites: [all]
contributions:
  authorship:
    - arash77
---

A growing number of Galaxy tools need to present pre-downloaded Hugging Face-compatible models to users at job submission time.  To avoid schema drift and duplicated administrator effort, the Galaxy community is standardising a **shared Galaxy data table** named `huggingface` that all such tools can consume from a single admin-managed `.loc` file.

[Flux](https://github.com/bgruening/galaxytools/tree/master/tools/flux) is the first tool to adopt this pattern, filtering models by type so users only see entries relevant to their task.

**Key points:**

- The table is named `huggingface` and lives in the standard Galaxy tool data directory
- It has **7 columns**: `value`, `name`, `pipeline_tag`, `domain`, `free_tag`, `version`, `path`
- Tool authors filter primarily by `pipeline_tag` and/or `domain`; `free_tag` is a fallback for finer-grained selection
- Rows are only added, never removed, so existing tool configurations stay stable as new models are registered
- A single `huggingface.loc` file is shared across all tools and maintained by admins

> The full schema reference, column conventions, and XML filter examples are being added to the [Galaxy admin documentation](https://docs.galaxyproject.org/en/master/admin/data_tables.html).

Author: Should I point it to dev?
```suggestion
The full schema reference, column conventions, and XML filter examples are being added to the [Galaxy admin documentation](https://docs.galaxyproject.org/en/latest/admin/data_tables.html).
```
